src/main/java/hudson/plugins/ec2/util/EC2SlaveConfig.java

```diff
@@ -0,0 +1,285 @@
+package hudson.plugins.ec2.util;
+
+import hudson.model.Node;
+import hudson.plugins.ec2.AMITypeData;
+import hudson.plugins.ec2.ConnectionStrategy;
+import hudson.plugins.ec2.EC2Tag;
+import hudson.slaves.NodeProperty;
+
+import java.util.List;
+
+public abstract class EC2SlaveConfig {
+
+    final String name;
+    final String description;
+    final String remoteFS;
+    final int numExecutors;
+    final String labelString;
+    final Node.Mode mode;
+    final String initScript;
+    final String tmpDir;
+    final List<? extends NodeProperty<?>> nodeProperties;
+    final String remoteAdmin;
+    final String jvmopts;
+    final String idleTerminationMinutes;
+    final List<EC2Tag> tags;
+    final String cloudName;
+    final int launchTimeout;
+    final AMITypeData amiType;
+    final ConnectionStrategy connectionStrategy;
+    final int maxTotalUses;
+
+    private EC2SlaveConfig(Builder<? extends Builder, ? extends EC2SlaveConfig> builder) {
+        this.name = builder.name;
+        this.description = builder.description;
+        this.remoteFS = builder.remoteFS;
+        this.numExecutors = builder.numExecutors;
+        this.labelString = builder.labelString;
+        this.mode = builder.mode;
+        this.initScript = builder.initScript;
+        this.tmpDir = builder.tmpDir;
+        this.nodeProperties = builder.nodeProperties;
+        this.remoteAdmin = builder.remoteAdmin;
+        this.jvmopts = builder.jvmopts;
+        this.idleTerminationMinutes = builder.idleTerminationMinutes;
+        this.tags = builder.tags;
+        this.cloudName = builder.cloudName;
+        this.launchTimeout = builder.launchTimeout;
+        this.amiType = builder.amiType;
+        this.connectionStrategy = builder.connectionStrategy;
+        this.maxTotalUses = builder.maxTotalUses;
+    }
+
+    public static class OnDemand extends EC2SlaveConfig {
+
+        final String instanceId;
+        final boolean stopOnTerminate;
+        final String publicDNS;
+        final String privateDNS;
+        final boolean useDedicatedTenancy;
+
+        private OnDemand(OnDemandBuilder builder) {
+            super(builder);
+            this.instanceId = builder.getInstanceId();
+            this.stopOnTerminate = builder.isStopOnTerminate();
+            this.publicDNS = builder.getPublicDNS();
+            this.privateDNS = builder.getPrivateDNS();
+            this.useDedicatedTenancy = builder.isUseDedicatedTenancy();
+        }
+    }
+
+    public static class Spot extends EC2SlaveConfig {
+
+        final String spotInstanceRequestId;
+
+        private Spot(SpotBuilder builder) {
+            super(builder);
+            this.spotInstanceRequestId = builder.spotInstanceRequestId;
+        }
+    }
+
+    private static abstract class Builder<B extends Builder, C extends EC2SlaveConfig> {
```

Comment:
```suggestion
    private static abstract class Builder<B extends Builder<B, C>, C extends EC2SlaveConfig> {
```

I think this would be a more appropriate generic bound here provided it still compiled. :)